Bugfix/rm 9504 tvp verursachen schlechte query plans

martin-herwich opened the pull request:
https://jira.rubicon.eu/browse/RM-9504

martin-herwich committed:
RM-9504 QueryDbCommandBuilder: generalize text replacement of query parameters

martin-herwich commented:
@MichaelKetting Der Stand vor dem letzten fixup-Commit ist der mit dem failing Test wegen Text-Parameter.

MichaelKetting commented on Remotion/Data/DomainObjects/Persistence/Rdbms/DbCommandBuilders/QueryDbCommandBuilder.cs:
```diff
@@ -28,51 +31,67 @@ namespace Remotion.Data.DomainObjects.Persistence.Rdbms.DbCommandBuilders
+      var statementBuilder = new StringBuilder(Statement.Length);
+      var statementIndex = 0;
 
-          statement = statement.Replace(parameterWithDefinition.QueryParameter.Name, queryParameterValueAsString);
```
okay, existing solution was a bit...iffy beceause it didn't account for parameter-name overlap. I would say that this isn't a feature but a bug and nobdy ever noticed it or there would be a unit test to confirm this behavior.

MichaelKetting replied:
To make matters worse, for us and our design, there's no naming rule for how the parameter name works with text parameters.

MichaelKetting replied:
Lastly, based on available usage information, we do have one user of the Text parameter, there it is used correctly with `@` prefixed.

MichaelKetting commented on Remotion/Data/DomainObjects/Persistence/Rdbms/DbCommandBuilders/QueryDbCommandBuilder.cs:
```diff
@@ -28,51 +31,67 @@ namespace Remotion.Data.DomainObjects.Persistence.Rdbms.DbCommandBuilders
   /// </summary>
   public class QueryDbCommandBuilder : DbCommandBuilder
```
Design:

notation for parameter placeholder in SQL Queries when using ADO.NET:
SQL Server: `@`
PostgreSQL: `$` or `@`
Oracle: `:`
MySql: `@`

So, basically, all systems we might consider relevant RDBMS use a prefix noation. This means we can add a test-method to ISqlDialect to check if a parameter name is valid.

Next, we can inject the sqldialect into the QueryDbCommandBuilder and make sure all parameters use a valid syntax.

Finally, this enables us to do a simple regex that checks for the parameters. We probably want to get the prefix-char from the sqldialect and then cache the regex in a static field for best performance.
for `@`: `\@\b[\w\d_]+\b`

This still finds usages of `@stuff` inside text, but only if the match also fits a know parameter would we actually have a false positive. this cannot be prevented and is an acceptable risk.

MichaelKetting commented on Remotion/Data/DomainObjects/Persistence/Rdbms/DbCommandBuilders/QueryDbCommandBuilder.cs:
```diff
@@ -28,51 +31,72 @@ namespace Remotion.Data.DomainObjects.Persistence.Rdbms.DbCommandBuilders
-      foreach (var parameterWithDefinition in _parametersWithDefinition)
-      {
-        var queryParameterValue = parameterWithDefinition.QueryParameter.Value;
+      var statementBuilder = new StringBuilder(Statement.Length);
```
we should allocate a string that's Length + Min(3*parameter-count, 100). Or just, Length+100... 

That way, we don't need to re-allocate if we do have to replace, but only waste 100 bytes of memory if we don't need a stringbuilder.

MichaelKetting commented on Remotion/Data/DomainObjects/Persistence/Rdbms/DbCommandBuilders/QueryDbCommandBuilder.cs:
```diff
@@ -28,51 +31,72 @@ namespace Remotion.Data.DomainObjects.Persistence.Rdbms.DbCommandBuilders
+      var regexKey = SqlDialect.GetParameterName("regex");
+      var paramRegex = s_regexDictionary.GetOrAdd(
+          regexKey,
+          key => new Regex(Regex.Escape(key).Replace("regex", "\\b[\\w\\d_]+\\b"), RegexOptions.IgnoreCase | RegexOptions.Multiline));
```
`Options.Compile`, please. and we don't need multiline because a parameter can't linewrap anyway.

MichaelKetting replied:
`Options.NoBacktracking` should also be helpful for performance
https://devblogs.microsoft.com/dotnet/regular-expression-improvements-in-dotnet-7/#backtracking-and-regexoptions-nonbacktracking

MichaelKetting commented on Remotion/Data/DomainObjects/Persistence/Rdbms/DbCommandBuilders/QueryDbCommandBuilder.cs:
```diff
@@ -28,51 +31,72 @@ namespace Remotion.Data.DomainObjects.Persistence.Rdbms.DbCommandBuilders
+      {
+        statementBuilder.Append(Statement.Substring(statementIndex, match.Index - statementIndex));
+        statementIndex = match.Index + match.Length;
+        var param = ParametersWithDefinition.SingleOrDefault(p => p.QueryParameter.Name.Equals(match.Value, StringComparison.OrdinalIgnoreCase));
```
In case of there being a 1000 parameters, this might be disadvantagous. We should dump the parmaters into a dictioanry. perhaps directly in the ctor.

MichaelKetting commented on Remotion/Data/DomainObjects/Persistence/Rdbms/DbCommandBuilders/QueryDbCommandBuilder.cs:
```diff
@@ -28,51 +31,72 @@ namespace Remotion.Data.DomainObjects.Persistence.Rdbms.DbCommandBuilders
+      var queryParameterValue = parameterWithDefinition.QueryParameter.Value;
+      if (parameterWithDefinition.QueryParameter.ParameterType == QueryParameterType.Text)
+      {
+        Assertion.DebugAssert(
+            queryParameterValue is string,
+            "parameterWithType.QueryParameter.Value is string when parameterWithType.QueryParameter.ParameterType == Text");
+        return (string)queryParameterValue;
```
I haven't thought this through, but am I seeing this correct that originally (and in the new code, of course), we only check that the value is a string for Debug builds? Which means consumers won't have this guard.

This feels like something we should change in a separate PR for v8.0 to a regular old InvalidOpEx. Thoughts?

martin-herwich replied:
I agree that this guard is only executed in Debug builds. I think that is because it helps find bugs during development, but is of little value in a production environment, since the type-cast in line 96 would fail anyways.
The guard provides a better error message during development, but would only waste CPU cycles in production.

MichaelKetting replied:
Aaah, true. thanks!

MichaelKetting commented on Remotion/Data/DomainObjects/Persistence/Rdbms/DbCommandBuilders/QueryDbCommandBuilder.cs:
```diff
@@ -28,51 +31,72 @@ namespace Remotion.Data.DomainObjects.Persistence.Rdbms.DbCommandBuilders
+            queryParameterValue is string,
+            "parameterWithType.QueryParameter.Value is string when parameterWithType.QueryParameter.ParameterType == Text");
+        return (string)queryParameterValue;
+      }
+
+      return parameterWithDefinition.QueryParameter.Name;
```
it feels like we should either switch this would look better using IF/ELSE because both branches are equally important.

MichaelKetting commented on Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/DbCommandBuilders/SqlDbQueryCommandBuilder.cs:
```diff
@@ -0,0 +1,104 @@
+using System;
```
lic header

MichaelKetting commented on Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/DbCommandBuilders/SqlDbQueryCommandBuilder.cs:
```diff
@@ -0,0 +1,104 @@
+
+namespace Remotion.Data.DomainObjects.Persistence.Rdbms.SqlServer.DbCommandBuilders;
+
+public class SqlDbQueryCommandBuilder : QueryDbCommandBuilder
```
xml doc

MichaelKetting commented on Remotion/Data/DomainObjects/Persistence/Rdbms/DbCommandBuilders/QueryDbCommandBuilder.cs:
```diff
@@ -28,51 +31,72 @@ namespace Remotion.Data.DomainObjects.Persistence.Rdbms.DbCommandBuilders
-      _statement = statement;
-      _parametersWithDefinition = parameters.ToArray();
+      Statement = statement;
+      ParametersWithDefinition = parameters.ToArray();
```
I would suggest calling it "paramerters". also, question is if a protected value is needed for statement and Parameters. 

given the later dictionary remark (whcih means IReadonlyDictionary), immuatiblity is already covered. We should have this immutable.

martin-herwich replied:
I am unsure what you mean with "question is if a protected value is needed". Do you mean that we should make these immutable properties public?

MichaelKetting replied:
Or just keep them private. Especially the statement will become obsolete during processing. And you can keep the parameters via the ctor if you need them. Which is only the case with the TVP because we want to proactively render the temp tables.

MichaelKetting commented on Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/DbCommandBuilders/SqlDbQueryCommandBuilder.cs:
```diff
@@ -0,0 +1,104 @@
+      return false;
+
+    var storageTypeLength = simpleStoragePropertyDefinition.ColumnDefinition.StorageTypeInfo.StorageTypeLength;
+    if(!storageTypeLength.HasValue)
```
```suggestion
    if (!storageTypeLength.HasValue)
```

MichaelKetting commented on Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/DbCommandBuilders/SqlDbQueryCommandBuilder.cs:
```diff
@@ -0,0 +1,104 @@
+
+    var storageDbType = simpleStoragePropertyDefinition.ColumnDefinition.StorageTypeInfo.StorageDbType;
+    var maxLength = storageDbType == DbType.String || storageDbType == DbType.StringFixedLength
+        ? 1700 / 2
+        : 1700;
```
plsase name the magic number

MichaelKetting commented on Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/DbCommandBuilders/SqlDbQueryCommandBuilder.cs:
```diff
@@ -0,0 +1,104 @@
+  }
+
+  public override IDbCommand Create (IDbCommandFactory dbCommandFactory)
+  {
```
nullcheck

MichaelKetting commented on Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/DbCommandBuilders/SqlDbQueryCommandBuilder.cs:
```diff
@@ -0,0 +1,104 @@
+
+  public override IDbCommand Create (IDbCommandFactory dbCommandFactory)
+  {
+    var command = base.Create(dbCommandFactory);
```
we should special case this. if there's no tvp, return the command.

MichaelKetting commented on Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/DbCommandBuilders/SqlDbQueryCommandBuilder.cs:
```diff
@@ -0,0 +1,104 @@
+
+    foreach (var tvp in _tvpToTempTable)
+    {
+      commandTextBuilder.AppendLine($"SELECT [Value] INTO {tvp.Value.TempTableName} FROM {tvp.Key};");
```
we use SqlDialect.DelimitIdentifier also in SqlTableValuedParameterComparedColumnSpecification. I can't say if that's just the only one where it's done that way or if this is the only possible place anyway. It just feels strange not doing it, but that's just my gut feeling

MichaelKetting commented on Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/DbCommandBuilders/SqlDbQueryCommandBuilder.cs:
```diff
@@ -0,0 +1,104 @@
+
+    foreach (var tvp in _tvpToTempTable)
+    {
+      commandTextBuilder.AppendLine($"SELECT [Value] INTO {tvp.Value.TempTableName} FROM {tvp.Key};");
+      if (tvp.Value.HasIndex)
+        commandTextBuilder.AppendLine($"CREATE NONCLUSTERED INDEX {tvp.Value.IndexName} ON {tvp.Value.TempTableName} ([Value]);");
```
I wonder if it wouldn't be better to use the brackets inline instead of in the replacement values. that way, we see that we are using them and where instead of seeing a mixe sql code. if we hardcode the brackets, it won't matter perfwise.

martin-herwich replied:
Instead of hard-coding the brackets, I now consistently use `SqlDialect.DelimitIdentifier()` - what do you think?

MichaelKetting replied:
Looks fine to me :)

MichaelKetting commented on Remotion/Data/DomainObjects.UnitTests/Persistence/Rdbms/DbCommandBuilders/QueryDbCommandBuilderTest.cs:
```diff
@@ -28,32 +32,93 @@ namespace Remotion.Data.DomainObjects.UnitTests.Persistence.Rdbms.DbCommandBuild
     [SetUp]
     public void SetUp ()
     {
       _valueParameterDefinition = new Mock<IDataParameterDefinition>(MockBehavior.Strict);
       _textParameterDefinition = new Mock<IDataParameterDefinition>(MockBehavior.Strict);
```
since 99% is new anyway: please rename to `Mock` suffixed

MichaelKetting commented on Remotion/Data/DomainObjects.UnitTests/Persistence/Rdbms/DbCommandBuilders/QueryDbCommandBuilderTest.cs:
```diff
@@ -66,21 +131,100 @@ public void Create ()
-          .Setup(mock => mock.CreateDataParameter(dbCommandStub.Object, _valueQueryParameter.Name, 5))
+
+      _valueParameterDefinition.Setup(mock => mock.GetParameterValue(5)).Returns(5);
+      _valueParameterDefinition.Setup(mock => mock.CreateDataParameter(dbCommandStub.Object, _valueQueryParameter.Name, _valueQueryParameter.Value))
```
```suggestion
      _valueParameterDefinition.Setup(mock => mock.CreateDataParameter(dbCommandStub.Object, _valueQueryParameter.Name, 5))
```
Keeping this as `5` makes it easier to reason over.

MichaelKetting commented on Remotion/Data/DomainObjects.UnitTests/Persistence/Rdbms/SqlServer/DbCommandBuilders/SqlQueryDbCommandBuilderTests.cs:
```diff
@@ -0,0 +1,157 @@
+using System;
```
lic header

MichaelKetting commented on Remotion/Data/DomainObjects.UnitTests/Persistence/Rdbms/SqlServer/DbCommandBuilders/SqlQueryDbCommandBuilderTests.cs:
```diff
@@ -0,0 +1,157 @@
+
+    var result = commandBuilder.Create(commandFactoryStub.Object);
+
+    Assert.That(
+        result.CommandText,
+        Is.EqualTo("""
+                   SELECT [Value] INTO [#@tvp] FROM @tvp;
+                   CREATE NONCLUSTERED INDEX [IX_@tvp] ON [#@tvp] ([Value]);
+
+                   SELECT * FROM [MyTable] WHERE MyRefID IN (SELECT [Value] FROM [#@tvp])
+
+                   DROP TABLE [#@tvp];
+
+                   """));
```
loopy stuff feels like a case where we should test the zero-items and the more-than-one-items cases. the zeri-items case might be overkill when testing, although it's also rather simple.

And we should test that we only do this for tvp parameters.

martin-herwich committed:
RM-9504 QueryDbCommandBuilder: change ctor parameter to IReadOnlyCollection

martin-herwich committed:
RM-9504 Use the new SqlQueryDbCommandBuilder

martin-herwich committed:
RM-9504 new SqlQueryDbCommandBuilder rewrites TVP usages

Each table-valued parameter is selected into a temp table (with index, if possible).
Usages of each TVP are rewritten to use the temp table instead.
